File: 2022/php/06.php
<?php

$string = file_get_contents('../input/06.txt');

const LAST_4 = 4;
const LAST_14 = 14;

$lastFour = [];

foreach (str_split($string) as $index => $character) {
    $lastFour[] = $character;

    if (count($lastFour) > LAST_14) array_shift($lastFour);

    if (arrayHasDupes($lastFour)) continue;

    if (count($lastFour) == LAST_14) {
        var_dump($index + 1);
        die();
    }
}

function arrayHasDupes(array $array) : bool {
    return count(array_unique($array)) !== count($array);
}

File: 2022/go/06.go
package main

import (
	"fmt"
	"os"
)

const PART_ONE = 4
const PART_TWO = 14

func main() {
bytes, _ := os.ReadFile("2022/input/06.txt")
	signal := string(bytes)

	var lastFour []rune

	for index, character := range signal {
		unique := make(map[rune]bool, 26)
		lastFour = append(lastFour, character)

		if len(lastFour) > PART_ONE {
			lastFour = lastFour[1:]
		}

		lastFourAreUnique := true
		for _, char := range lastFour {

			if unique[char] {
				lastFourAreUnique = false
				break
			}

			unique[char] = true
		}

		if lastFourAreUnique && len(lastFour) == PART_ONE {
			fmt.Println(index + 1)
			break
		}
	}
}
